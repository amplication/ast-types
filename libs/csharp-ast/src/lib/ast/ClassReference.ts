import { AstNode } from "../core/AstNode";
import { Writer } from "../core/Writer";

/**
 * Configuration arguments for creating a C# class reference.
 */
export declare namespace ClassReference {
  interface Args {
    /** The name of the C# class */
    name: string;
    /** The namespace of the C# class */
    namespace: string;
  }
}

/**
 * Represents a reference to a C# class in the AST.
 * This class handles the generation of C# class references including
 * the class name and its namespace.
 *
 * @extends {AstNode}
 */
export class ClassReference extends AstNode {
  /** The name of the class */
  public readonly name: string;
  /** The namespace containing the class */
  public readonly namespace: string;

  /**
   * Creates a new C# class reference.
   * @param {ClassReference.Args} args - The configuration arguments for the class reference
   */
  constructor({ name, namespace }: ClassReference.Args) {
    super();
    this.name = name;
    this.namespace = namespace;
  }

  /**
   * Writes the class reference to the writer.
   * @param {Writer} writer - The writer to write to
   */
  public write(writer: Writer): void {
    writer.addReference(this);
    writer.write(`${this.name}`);
  }
}

/**
 * Pre-defined class reference for the OneOf type.
 */
export const OneOfClassReference = new ClassReference({
  name: "OneOf",
  namespace: "OneOf",
});

//   TODO: remove this in favor of the one in PrebuiltUtilities

/**
 * Pre-defined class reference for the StringEnum type.
 */
export const StringEnumClassReference = new ClassReference({
  name: "StringEnum",
  namespace: "StringEnum",
});
