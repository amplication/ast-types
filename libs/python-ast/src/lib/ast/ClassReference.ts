import { AstNode } from "../core/AstNode";
import { Writer } from "../core/Writer";

/**
 * Configuration arguments for creating a Python class reference.
 */
export interface ClassReferenceArgs {
  /** The name of the class */
  name: string;
  /** The module name containing the class */
  moduleName?: string;
  /** Optional alias for the class when imported */
  alias?: string;
}

/**
 * Represents a reference to a Python class.
 * This class is used to track and generate imports for classes used in the code.
 *
 * @extends {AstNode}
 */
export class ClassReference extends AstNode {
  /** The name of the class */
  public readonly name: string;
  /** The module name containing the class */
  public readonly moduleName?: string;
  /** Optional alias for the class */
  public readonly alias?: string;

  /**
   * Creates a new Python class reference.
   * @param {ClassReferenceArgs} args - The configuration arguments
   */
  constructor({ name, moduleName, alias }: ClassReferenceArgs) {
    super();
    this.name = name;
    this.moduleName = moduleName;
    this.alias = alias;
  }

  /**
   * Writes the class reference to the writer and adds it to the import collection.
   * @param {Writer} writer - The writer to write to
   */
  public write(writer: Writer): void {
    if (this.moduleName) {
      writer.addImport(this);
    }
    writer.write(this.alias || this.name);
  }
}
