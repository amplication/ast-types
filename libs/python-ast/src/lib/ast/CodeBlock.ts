import { AstNode } from "../core/AstNode";
import { Writer } from "../core/Writer";
import { Import } from "./Import";

/**
 * Configuration arguments for creating a Python code block.
 */
export type CodeBlockArgs =
  | {
      /** The code to write */
      code: string;
      /** A list of imports that are used in the code */
      imports?: Import[] | null;
    }
  | {
      /** A function that writes code to the writer */
      code: (writer: Writer) => void;
      imports?: never;
    };

/**
 * Represents a block of Python code in the AST.
 * This class handles the generation of arbitrary Python code blocks, either as
 * static strings or as dynamic code generation functions. It is used for language
 * features not directly supported by specific AST node classes.
 *
 * @extends {AstNode}
 */
export class CodeBlock extends AstNode {
  /** The code content or code generation function */
  private value: string | ((writer: Writer) => void);
  /** The imports used in the code */
  private imports: Import[];

  /**
   * Creates a new Python code block.
   * @param {CodeBlockArgs} args - The configuration arguments
   */
  constructor(args: CodeBlockArgs) {
    super();
    this.value = args.code;
    this.imports = [];

    if ("imports" in args && args.imports) {
      this.imports.push(...args.imports);
    }
  }

  /**
   * Writes the code block to the writer.
   * @param {Writer} writer - The writer to write to
   */
  public write(writer: Writer): void {
    if (typeof this.value === "string") {
      this.imports.forEach((import_) => writer.addImport(import_));
      writer.write(this.value);
    } else {
      this.value(writer);
    }
  }
}
