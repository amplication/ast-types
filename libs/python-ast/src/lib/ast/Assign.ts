import { AstNode } from "../core/AstNode";
import { Writer } from "../core/Writer";
import { ClassReference } from "./ClassReference";

/**
 * Configuration arguments for creating a Python assignment statement.
 */
export interface AssignArgs {
  /** The target variable name */
  target: string;
  /** The value to assign */
  value: string;
  /** Type annotation for the target */
  type?: ClassReference;
  /** Whether this is a class variable (vs. instance variable) */
  isClassVariable?: boolean;
}

/**
 * Represents a Python assignment statement.
 * This class handles the generation of Python assignment statements, which can
 * include type annotations and handle class vs. instance variables.
 *
 * @extends {AstNode}
 */
export class Assign extends AstNode {
  /** The target variable name */
  public readonly target: string;
  /** The value to assign */
  public readonly value: string;
  /** Type annotation for the target */
  public readonly type?: ClassReference;
  /** Whether this is a class variable */
  public readonly isClassVariable: boolean;

  /**
   * Creates a new Python assignment statement.
   * @param {AssignArgs} args - The configuration arguments
   */
  constructor({ target, value, type, isClassVariable = false }: AssignArgs) {
    super();
    this.target = target;
    this.value = value;
    this.type = type;
    this.isClassVariable = isClassVariable;
  }

  /**
   * Writes the assignment statement to the writer.
   * @param {Writer} writer - The writer to write to
   */
  public write(writer: Writer): void {
    writer.write(this.target);

    if (this.type) {
      writer.write(": ");
      this.type.write(writer);
      writer.write(" = ");
    } else {
      writer.write(" = ");
    }

    writer.write(this.value);
  }
}
