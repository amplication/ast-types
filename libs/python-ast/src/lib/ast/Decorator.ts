import { AstNode } from "../core/AstNode";
import { Writer } from "../core/Writer";
import { ClassReference } from "./ClassReference";

/**
 * Configuration arguments for creating a Python decorator.
 */
export interface DecoratorArgs {
  /** The name of the decorator */
  name: string;
  /** The arguments to pass to the decorator */
  args?: string[];
  /** The keyword arguments to pass to the decorator */
  kwargs?: Record<string, string>;
  /** Optional module name for import */
  moduleName?: string;
}

/**
 * Represents a Python decorator.
 * This class handles the generation of Python decorators, which are used to modify
 * the behavior of functions, methods, and classes.
 *
 * @extends {AstNode}
 */
export class Decorator extends AstNode {
  /** The name of the decorator */
  public readonly name: string;
  /** The arguments to pass to the decorator */
  public readonly args: string[];
  /** The keyword arguments to pass to the decorator */
  public readonly kwargs: Record<string, string>;
  /** The module name containing the decorator */
  public readonly moduleName?: string;
  /** The class reference for import handling */
  private readonly reference?: ClassReference;

  /**
   * Creates a new Python decorator.
   * @param {DecoratorArgs} args - The configuration arguments
   */
  constructor({ name, args = [], kwargs = {}, moduleName }: DecoratorArgs) {
    super();
    this.name = name;
    this.args = args;
    this.kwargs = kwargs;
    this.moduleName = moduleName;

    if (moduleName) {
      this.reference = new ClassReference({
        name: this.name,
        moduleName: this.moduleName,
      });
    }
  }

  /**
   * Writes the decorator to the writer.
   * @param {Writer} writer - The writer to write to
   */
  public write(writer: Writer): void {
    writer.write("@");

    if (this.reference) {
      this.reference.write(writer);
    } else {
      writer.write(this.name);
    }

    // Add arguments and keyword arguments if provided
    const hasArgs = this.args.length > 0;
    const hasKwargs = Object.keys(this.kwargs).length > 0;

    if (hasArgs || hasKwargs) {
      writer.write("(");

      // Write positional arguments
      this.args.forEach((arg, index) => {
        writer.write(arg);
        if (index < this.args.length - 1 || hasKwargs) {
          writer.write(", ");
        }
      });

      // Write keyword arguments
      Object.entries(this.kwargs).forEach(([key, value], index, array) => {
        writer.write(`${key}=${value}`);
        if (index < array.length - 1) {
          writer.write(", ");
        }
      });

      writer.write(")");
    }
  }
}
