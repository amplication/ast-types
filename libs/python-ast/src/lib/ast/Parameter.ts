import { AstNode } from "../core/AstNode";
import { Writer } from "../core/Writer";
import { ClassReference } from "./ClassReference";

/**
 * Configuration arguments for creating a Python parameter.
 */
export interface ParameterArgs {
  /** The name of the parameter */
  name: string;
  /** The type annotation for the parameter */
  type?: ClassReference;
  /** The default value for the parameter */
  default_?: string;
  /** Whether this is a keyword-only parameter (after *) */
  isKeywordOnly?: boolean;
  /** Whether this is a positional-only parameter (before /) */
  isPositionalOnly?: boolean;
  /** Whether this is a variable positional parameter (*args) */
  isVariablePositional?: boolean;
  /** Whether this is a variable keyword parameter (**kwargs) */
  isVariableKeyword?: boolean;
}

/**
 * Represents a Python parameter in a function or method definition.
 * This class handles the generation of parameter declarations including
 * type annotations and default values.
 *
 * @extends {AstNode}
 */
export class Parameter extends AstNode {
  /** The name of the parameter */
  public readonly name: string;
  /** The type annotation for the parameter */
  public readonly type?: ClassReference;
  /** The default value for the parameter */
  public readonly default_?: string;
  /** Whether this is a keyword-only parameter */
  public readonly isKeywordOnly: boolean;
  /** Whether this is a positional-only parameter */
  public readonly isPositionalOnly: boolean;
  /** Whether this is a variable positional parameter */
  public readonly isVariablePositional: boolean;
  /** Whether this is a variable keyword parameter */
  public readonly isVariableKeyword: boolean;

  /**
   * Creates a new Python parameter.
   * @param {ParameterArgs} args - The configuration arguments
   */
  constructor({
    name,
    type,
    default_,
    isKeywordOnly = false,
    isPositionalOnly = false,
    isVariablePositional = false,
    isVariableKeyword = false,
  }: ParameterArgs) {
    super();
    this.name = name;
    this.type = type;
    this.default_ = default_;
    this.isKeywordOnly = isKeywordOnly;
    this.isPositionalOnly = isPositionalOnly;
    this.isVariablePositional = isVariablePositional;
    this.isVariableKeyword = isVariableKeyword;

    // Validate parameter configuration
    if (
      [
        isKeywordOnly,
        isPositionalOnly,
        isVariablePositional,
        isVariableKeyword,
      ].filter(Boolean).length > 1
    ) {
      throw new Error(
        "A parameter can only be one of: keyword-only, positional-only, variable positional, or variable keyword",
      );
    }
  }

  /**
   * Writes the parameter declaration to the writer.
   * @param {Writer} writer - The writer to write to
   */
  public write(writer: Writer): void {
    if (this.isVariablePositional) {
      writer.write("*");
    } else if (this.isVariableKeyword) {
      writer.write("**");
    }

    writer.write(this.name);

    if (this.type) {
      writer.write(": ");
      this.type.write(writer);
    }

    if (this.default_) {
      writer.write(" = ");
      writer.write(this.default_);
    }
  }
}
