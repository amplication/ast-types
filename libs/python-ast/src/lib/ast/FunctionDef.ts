import { AstNode } from "../core/AstNode";
import { Writer } from "../core/Writer";
import { ClassReference } from "./ClassReference";
import { CodeBlock } from "./CodeBlock";
import { Decorator } from "./Decorator";
import { Parameter } from "./Parameter";
import { Return } from "./Return";

/**
 * Configuration arguments for creating a Python function.
 */
export interface FunctionDefArgs {
  /** The name of the function */
  name: string;
  /** The parameters of the function */
  parameters?: Parameter[];
  /** Decorators to apply to the function */
  decorators?: Decorator[];
  /** Documentation string for the function */
  docstring?: string;
  /** Return type annotation */
  returnType?: ClassReference;
  /** Whether this is a static method */
  isStatic?: boolean;
  /** Whether this is a class method */
  isClassMethod?: boolean;
  /** Whether this is an async function */
  isAsync?: boolean;
}

/**
 * Represents a Python function definition in the AST.
 * This class handles the generation of Python function declarations including
 * parameters, decorators, and return type annotations.
 *
 * @extends {AstNode}
 */
export class FunctionDef extends AstNode {
  /** The name of the function */
  public readonly name: string;
  /** The parameters of the function */
  public readonly parameters: Parameter[];
  /** Decorators applied to the function */
  public readonly decorators: Decorator[];
  /** The documentation string */
  public readonly docstring?: string;
  /** The return type annotation */
  public readonly returnType?: ClassReference;
  /** Whether this is a static method */
  public readonly isStatic: boolean;
  /** Whether this is a class method */
  public readonly isClassMethod: boolean;
  /** Whether this is an async function */
  public readonly isAsync: boolean;

  /** The body of the function */
  private body: (CodeBlock | Return)[] = [];

  /**
   * Creates a new Python function definition.
   * @param {FunctionDefArgs} args - The configuration arguments
   */
  constructor({
    name,
    parameters = [],
    decorators = [],
    docstring,
    returnType,
    isStatic = false,
    isClassMethod = false,
    isAsync = false,
  }: FunctionDefArgs) {
    super();
    this.name = name;
    this.parameters = parameters;
    this.decorators = [...decorators];
    this.docstring = docstring;
    this.returnType = returnType;
    this.isStatic = isStatic;
    this.isClassMethod = isClassMethod;
    this.isAsync = isAsync;

    // Add appropriate decorators for static and class methods
    if (
      this.isStatic &&
      !this.decorators.some((d) => d.name === "staticmethod")
    ) {
      this.decorators.push(new Decorator({ name: "staticmethod" }));
    }
    if (
      this.isClassMethod &&
      !this.decorators.some((d) => d.name === "classmethod")
    ) {
      this.decorators.push(new Decorator({ name: "classmethod" }));
    }
  }

  /**
   * Adds a statement to the function body.
   * @param {CodeBlock | Return} statement - The statement to add
   */
  public addStatement(statement: CodeBlock | Return): void {
    this.body.push(statement);
  }

  /**
   * Writes the function definition and its body to the writer.
   * @param {Writer} writer - The writer to write to
   */
  public write(writer: Writer): void {
    // Write decorators
    this.decorators.forEach((decorator) => {
      decorator.write(writer);
      writer.newLine();
    });

    // Write function definition
    if (this.isAsync) {
      writer.write("async ");
    }
    writer.write(`def ${this.name}(`);

    // Write parameters
    this.parameters.forEach((param, index) => {
      param.write(writer);
      if (index < this.parameters.length - 1) {
        writer.write(", ");
      }
    });

    writer.write(")");

    // Write return type annotation if provided
    if (this.returnType) {
      writer.write(" -> ");
      this.returnType.write(writer);
    }

    writer.write(":");
    writer.newLine();
    writer.indent();

    // Write docstring if provided
    if (this.docstring) {
      writer.writeLine(`"""${this.docstring}"""`);
      if (this.body.length > 0) {
        writer.newLine();
      }
    }

    // Write function body
    if (this.body.length > 0) {
      this.body.forEach((statement) => {
        statement.write(writer);
        writer.newLine();
      });
    } else {
      // Empty function body needs a pass statement
      writer.writeLine("pass");
    }

    writer.dedent();
  }
}
