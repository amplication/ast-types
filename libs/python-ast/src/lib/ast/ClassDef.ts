import { AstNode } from "../core/AstNode";
import { Writer } from "../core/Writer";
import { Assign } from "./Assign";
import { ClassReference } from "./ClassReference";
import { CodeBlock } from "./CodeBlock";
import { Decorator } from "./Decorator";
import { FunctionDef } from "./FunctionDef";

/**
 * Configuration arguments for creating a Python class.
 */
export interface ClassDefArgs {
  /** The name of the class */
  name: string;
  /** The module name containing the class */
  moduleName?: string;
  /** The parent classes to inherit from */
  bases?: ClassReference[];
  /** Decorators to apply to the class */
  decorators?: Decorator[];
  /** Documentation string for the class */
  docstring?: string;
}

/**
 * Represents a Python class definition in the AST.
 * This class handles the generation of Python class declarations including methods,
 * attributes, and nested classes.
 *
 * @extends {AstNode}
 */
export class ClassDef extends AstNode {
  /** The name of the class */
  public readonly name: string;
  /** The module name containing the class */
  public readonly moduleName?: string;
  /** The parent classes to inherit from */
  public readonly bases: ClassReference[];
  /** The decorators applied to the class */
  public readonly decorators: Decorator[];
  /** The documentation string for the class */
  public readonly docstring?: string;
  /** The class reference for this class */
  public readonly reference: ClassReference;

  /** The methods defined in this class */
  private methods: FunctionDef[] = [];
  /** The attributes defined in this class */
  private attributes: Assign[] = [];
  /** The nested classes defined in this class */
  private nestedClasses: ClassDef[] = [];
  /** Custom code blocks for unsupported features */
  private codeBlocks: CodeBlock[] = [];

  /**
   * Creates a new Python class definition.
   * @param {ClassDefArgs} args - The configuration arguments
   */
  constructor({
    name,
    moduleName,
    bases = [],
    decorators = [],
    docstring,
  }: ClassDefArgs) {
    super();
    this.name = name;
    this.moduleName = moduleName;
    this.bases = bases;
    this.decorators = decorators;
    this.docstring = docstring;

    this.reference = new ClassReference({
      name: this.name,
      moduleName: this.moduleName,
    });
  }

  /**
   * Adds a method to the class.
   * @param {FunctionDef} method - The method to add
   */
  public addMethod(method: FunctionDef): void {
    this.methods.push(method);
  }

  /**
   * Adds an attribute to the class.
   * @param {Assign} attribute - The attribute to add
   */
  public addAttribute(attribute: Assign): void {
    this.attributes.push(attribute);
  }

  /**
   * Adds a nested class to this class.
   * @param {ClassDef} nestedClass - The nested class to add
   */
  public addNestedClass(nestedClass: ClassDef): void {
    this.nestedClasses.push(nestedClass);
  }

  /**
   * Adds a code block for unsupported features.
   * @param {CodeBlock} codeBlock - The code block to add
   */
  public addCodeBlock(codeBlock: CodeBlock): void {
    this.codeBlocks.push(codeBlock);
  }

  /**
   * Writes the class definition and its members to the writer.
   * @param {Writer} writer - The writer to write to
   */
  public write(writer: Writer): void {
    // Write decorators
    this.decorators.forEach((decorator) => {
      decorator.write(writer);
      writer.newLine();
    });

    // Write class definition
    writer.write(`class ${this.name}`);

    // Write base classes if any
    if (this.bases.length > 0) {
      writer.write("(");
      this.bases.forEach((base, index) => {
        base.write(writer);
        if (index < this.bases.length - 1) {
          writer.write(", ");
        }
      });
      writer.write(")");
    }

    writer.write(":");
    writer.newLine();
    writer.indent();

    // Write docstring if provided
    if (this.docstring) {
      writer.writeLine(`"""${this.docstring}"""`);
      if (
        this.attributes.length > 0 ||
        this.methods.length > 0 ||
        this.nestedClasses.length > 0 ||
        this.codeBlocks.length > 0
      ) {
        writer.newLine();
      }
    }

    // Write attributes
    if (this.attributes.length > 0) {
      this.attributes.forEach((attribute) => {
        attribute.write(writer);
        writer.newLine();
      });
      if (
        this.methods.length > 0 ||
        this.nestedClasses.length > 0 ||
        this.codeBlocks.length > 0
      ) {
        writer.newLine();
      }
    }

    // Write methods
    this.methods.forEach((method, index, array) => {
      method.write(writer);
      if (index < array.length - 1) {
        writer.newLine();
      }
    });

    if (
      this.methods.length > 0 &&
      (this.nestedClasses.length > 0 || this.codeBlocks.length > 0)
    ) {
      writer.newLine();
      writer.newLine();
    }

    // Write nested classes
    this.nestedClasses.forEach((nestedClass, index, array) => {
      nestedClass.write(writer);
      if (index < array.length - 1) {
        writer.newLine();
        writer.newLine();
      }
    });

    if (this.nestedClasses.length > 0 && this.codeBlocks.length > 0) {
      writer.newLine();
      writer.newLine();
    }

    // Write additional code blocks
    this.codeBlocks.forEach((codeBlock, index, array) => {
      codeBlock.write(writer);
      if (index < array.length - 1) {
        writer.newLine();
      }
    });

    // If the class is empty, add a pass statement
    if (
      this.attributes.length === 0 &&
      this.methods.length === 0 &&
      this.nestedClasses.length === 0 &&
      this.codeBlocks.length === 0 &&
      !this.docstring
    ) {
      writer.writeLine("pass");
    }

    writer.dedent();
  }
}
