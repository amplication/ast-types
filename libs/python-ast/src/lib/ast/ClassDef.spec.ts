import { ClassDef } from "./ClassDef";
import { FunctionDef } from "./FunctionDef";
import { Parameter } from "./Parameter";
import { ClassReference } from "./ClassReference";
import { CodeBlock } from "./CodeBlock";
import { Decorator } from "./Decorator";
import { Assign } from "./Assign";

describe("ClassDef", () => {
  it("should generate a simple class with docstring", () => {
    const classDef = new ClassDef({
      name: "Person",
      docstring: "A class representing a person",
    });

    expect(classDef.toString()).toMatchSnapshot();
    expect(classDef.toString()).toContain("class Person:");
    expect(classDef.toString()).toContain(
      '"""A class representing a person"""',
    );
  });

  it("should generate a class with inheritance", () => {
    const classDef = new ClassDef({
      name: "Student",
      bases: [new ClassReference({ name: "Person", moduleName: "models" })],
    });

    expect(classDef.toString()).toMatchSnapshot();
    expect(classDef.toString()).toContain("from models import Person");
    expect(classDef.toString()).toContain("class Student(Person):");
  });

  it("should generate a class with methods", () => {
    const classDef = new ClassDef({
      name: "Calculator",
    });

    const addMethod = new FunctionDef({
      name: "add",
      parameters: [
        new Parameter({ name: "self" }),
        new Parameter({ name: "a", type: new ClassReference({ name: "int" }) }),
        new Parameter({ name: "b", type: new ClassReference({ name: "int" }) }),
      ],
      returnType: new ClassReference({ name: "int" }),
    });

    addMethod.addStatement(new CodeBlock({ code: "return a + b" }));
    classDef.addMethod(addMethod);

    expect(classDef.toString()).toMatchSnapshot();
    expect(classDef.toString()).toContain(
      "def add(self, a: int, b: int) -> int:",
    );
    expect(classDef.toString()).toContain("return a + b");
  });

  it("should generate a class with attributes", () => {
    const classDef = new ClassDef({
      name: "Configuration",
    });

    classDef.addAttribute(
      new Assign({
        target: "DEFAULT_TIMEOUT",
        value: "30",
        isClassVariable: true,
      }),
    );

    classDef.addAttribute(
      new Assign({
        target: "DEBUG",
        value: "False",
        isClassVariable: true,
      }),
    );

    expect(classDef.toString()).toMatchSnapshot();
    expect(classDef.toString()).toContain("DEFAULT_TIMEOUT = 30");
    expect(classDef.toString()).toContain("DEBUG = False");
  });

  it("should generate a class with decorators", () => {
    const classDef = new ClassDef({
      name: "Singleton",
      decorators: [
        new Decorator({ name: "dataclass", moduleName: "dataclasses" }),
      ],
    });

    expect(classDef.toString()).toMatchSnapshot();
    expect(classDef.toString()).toContain("from dataclasses import dataclass");
    expect(classDef.toString()).toContain("@dataclass");
  });

  it("should generate a complete class with multiple features", () => {
    const classDef = new ClassDef({
      name: "User",
      docstring: "A class representing a user in the system",
      bases: [new ClassReference({ name: "BaseModel", moduleName: "models" })],
      decorators: [
        new Decorator({ name: "dataclass", moduleName: "dataclasses" }),
      ],
    });

    // Add class attributes
    classDef.addAttribute(
      new Assign({
        target: "table_name",
        value: "'users'",
        isClassVariable: true,
      }),
    );

    // Add __init__ method
    const initMethod = new FunctionDef({
      name: "__init__",
      parameters: [
        new Parameter({ name: "self" }),
        new Parameter({
          name: "username",
          type: new ClassReference({ name: "str" }),
        }),
        new Parameter({
          name: "email",
          type: new ClassReference({ name: "str" }),
        }),
        new Parameter({
          name: "is_active",
          type: new ClassReference({ name: "bool" }),
          defaultValue: "True",
        }),
      ],
    });
    initMethod.addStatement(
      new CodeBlock({
        code: "self.username = username\nself.email = email\nself.is_active = is_active",
      }),
    );
    classDef.addMethod(initMethod);

    // Add toString method
    const strMethod = new FunctionDef({
      name: "__str__",
      parameters: [new Parameter({ name: "self" })],
      returnType: new ClassReference({ name: "str" }),
    });
    strMethod.addStatement(
      new CodeBlock({
        code: "return f'User({self.username}, {self.email})'",
      }),
    );
    classDef.addMethod(strMethod);

    expect(classDef.toString()).toMatchSnapshot();
  });
});
