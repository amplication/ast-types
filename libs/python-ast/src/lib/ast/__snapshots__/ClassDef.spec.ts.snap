// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClassDef should generate a class with attributes 1`] = `
"class Configuration:
    DEFAULT_TIMEOUT = 30
    DEBUG = False
"
`;

exports[`ClassDef should generate a class with decorators 1`] = `
"from dataclasses import dataclass

@dataclass
class Singleton:
    pass
"
`;

exports[`ClassDef should generate a class with inheritance 1`] = `
"from models import Person

class Student(Person):
    pass
"
`;

exports[`ClassDef should generate a class with methods 1`] = `
"class Calculator:
    def add(self, a: int, b: int) -> int:
        return a + b
"
`;

exports[`ClassDef should generate a complete class with multiple features 1`] = `
"from dataclasses import dataclass
from models import BaseModel

@dataclass
class User(BaseModel):
    """A class representing a user in the system"""

    table_name = 'users'

    def __init__(self, username: str, email: str, is_active: bool = True):
        self.username = username
        self.email = email
        self.is_active = is_active

    def __str__(self) -> str:
        return f'User({self.username}, {self.email})'
"
`;

exports[`ClassDef should generate a simple class with docstring 1`] = `
"class Person:
    """A class representing a person"""
"
`;
